src/xdp/rx.c

```diff
@@ -1110,12 +1123,12 @@ XdpRxQueueSetProgram(
     return Status;
 }
 
-XDP_PROGRAM *
-XdpRxQueueGetProgram(
+LIST_ENTRY *
+XdpRxQueueGetProgramBindingList(
```

Comment:
IMO this abstraction is going to be fragile once we integrate with eBPF, because we'll want to invoke either eBPF or our legacy built-in programs, but not both. It would be best if the RX queue exposes an opaque `XDP_PROGRAM *` pointer getter/setter and then invokes the XDP_PROGRAM on the data path.

Reply:
Sorry, I am not quite following here. Are you saying this new concept of "program binding" would not exist with ebpf? Or this abstraction allows mixing ebpf and non-ebpf programs?

Reply:
Hmm, a notion of "program binding" definitely exists with eBPF, so the issue is the mixing of different types of programs, and to a lesser extent the leakage of the chain of non-eBPF programs into the RX queue. One option here is to create a top-level program binding (it could contain either an eBPF program or a list of non-eBPF programs) and then have the RX queue deal only with that.

For the non-eBPF programs, then the program.c module could peek into the list stored in that top-level binding object.

Reply:
I agree. That would be cleaner. Do you want me to work on that in this PR?

Reply:
(tests are all green now except a flaky functional test case)

Reply:
Why do we need to keep a legacy mechanism once ebpf integration is supported?

Reply:
We may be in a partially-integrated state for a while, where some features of our built-in program implementation are yet to be lit up through eBPF. Once everyone depending on those features can switch to eBPF, we should be able to migrate away from built-in programs.

Reply:
Sounds like you're suggesting two separate steps, i.e., first add ebpf integration, then remove built-in programs?  The word "replace" in the title of issue #7 implied to me that there is no "partially integrated state for a while", it's all at the same time.   I'm asking whether you actually need a partially integrated state for a while instead of just doing it.

Reply:
@dthaler MsQuic is dependent on the built-in program support. And we are going forward with deployments leveraging that. Until we have a complete replacement for those deployments, we cannot remove the built-in program support.